src/hivpy/prep.py

```diff
@@ -750,21 +783,47 @@ def restart_prep(self, pop: Population, time_step):
                 pop.set_present_variable(col.CONT_ACTIVE_ON_PREP, time_step, restarting_prep_pop)
                 # increment cumulative use
                 self.set_all_prep_cumulative(pop, restarting_prep_pop, time_step)
+                pop.set_present_variable(col.LAST_PREP_USE_DATE, pop.date, restarting_prep_pop)
                 # unset stop date
                 pop.set_present_variable(col.LAST_PREP_STOP_DATE, None, restarting_prep_pop)
+                # unpause prep
+                pop.set_present_variable(col.PREP_PAUSED, False, restarting_prep_pop)
 
-    def calc_restarting_prep(self, favoured_prep, size):
+    def calc_restarting_prep(self, favoured_prep, prep_paused, size):
         """
         Returns PrEP types for people restarting PrEP.
         Individual preferences and availability are taken into account.
         """
         # outcomes
         r = rng.uniform(size=size)
-        restarting = r < self.prob_prep_restart
+        # those who stopped due to ineligibility are guaranteed to restart
+        if prep_paused:
+            restarting = r < 1
+        else:
+            restarting = r < self.prob_prep_restart
         prep = [favoured_prep if r else None for r in restarting]
 
         return prep
 
+    def stop_prep(self, pop: Population):
+        """
+        Update PrEP usage for people stopping PrEP due to lack of eligibility or positive test.
+        """
+        # people who are using prep and are now ineligible
+        ineligible = pop.get_sub_pop(AND(OR(COND(col.HIV_DIAGNOSED, op.eq, True),
+                                            COND(col.PREP_ELIGIBLE, op.eq, False)),
+                                         COND(col.EVER_PREP, op.eq, True),
+                                         COND(col.LAST_PREP_STOP_DATE, op.eq, None),
+                                         COND(col.LAST_TEST_DATE, op.eq, pop.date)))
+
+        if len(ineligible) > 0:
+            # reset active continuous use
+            pop.set_present_variable(col.CONT_ACTIVE_ON_PREP, timedelta(months=0), ineligible)
+            # set stop date
+            pop.set_present_variable(col.LAST_PREP_STOP_DATE, pop.date, ineligible)
+            # pause prep
+            pop.set_present_variable(col.PREP_PAUSED, True, ineligible)
```

Comment:
Should people who are diagnosed -- and therefore going off PrEP forever -- be considered "paused"?

Reply:
A few things to change:
- Add a new column (something like `CONT_INTENT_ON_PREP`) and continue to increment it for people who have paused PrEP but still intend to be on PrEP
- Add permanent usage cutoffs, such as being diagnosed and aging out of PrEP (past this point people should no longer be paused, they should be off PrEP permanently)

Reply:
- People who are diagnosed need to be off prep rather than paused

Additional variable e.g. `CONT_PREP_INTENT`
- This would be time when they would take prep when eligible i.e. _including_ pauses
- This stops counts up if they are over 65 or diagnosed